python/dgl/ops/spmm.py

```diff
@@ -43,31 +68,30 @@ def gspmm(g, op, reduce_op, lhs_data, rhs_data):
     tensor
         The result tensor.
     """
-    if op not in ['copy_lhs', 'copy_rhs']:
-        # Expand dims so that there will be no broadcasting issues with different
-        # number of dimensions. For example, given two shapes (N, 3, 1), (E, 5, 3, 4)
-        # that are valid broadcastable shapes, change them to (N, 1, 3, 1) and
-        # (E, 5, 3, 4)
-        lhs_shape = F.shape(lhs_data)
-        rhs_shape = F.shape(rhs_data)
-        if len(lhs_shape) != len(rhs_shape):
-            max_ndims = max(len(lhs_shape), len(rhs_shape))
-            lhs_pad_ndims = max_ndims - len(lhs_shape)
-            rhs_pad_ndims = max_ndims - len(rhs_shape)
-            new_lhs_shape = (lhs_shape[0],) + (1,) * lhs_pad_ndims + lhs_shape[1:]
-            new_rhs_shape = (rhs_shape[0],) + (1,) * rhs_pad_ndims + rhs_shape[1:]
-            lhs_data = F.reshape(lhs_data, new_lhs_shape)
-            rhs_data = F.reshape(rhs_data, new_rhs_shape)
-    # With max and min reducers infinity will be returned for zero degree nodes
-    ret = gspmm_internal(g._graph, op,
-                         'sum' if reduce_op == 'mean' else reduce_op,
-                         lhs_data, rhs_data)
-    # Replace infinity with zero for isolated nodes when reducer is min/max
-    if reduce_op in ['min', 'max']:
-        ret = F.replace_inf_with_zero(ret)
+    if g._graph.number_of_etypes() == 1:
+        if op not in ['copy_lhs', 'copy_rhs']:
+            lhs_data, rhs_data = reshape_lhs_rhs(lhs_data, rhs_data)
+        # With max and min reducers infinity will be returned for zero degree nodes
+        ret = gspmm_internal(g._graph, op,
+                             'sum' if reduce_op == 'mean' else reduce_op,
+                             lhs_data, rhs_data)
+        # Replace infinity with zero for isolated nodes when reducer is min/max
+        if reduce_op in ['min', 'max']:
+            ret = F.replace_inf_with_zero(ret)
+    else:
+        if op in ['copy_lhs', 'copy_rhs']:
+            lhs_and_rhs_tuple = lhs_data if rhs_data is None else rhs_data
+        ret = gspmm_internal_hetero(g, op,
+                                    'sum' if reduce_op == 'mean' else reduce_op,
+                                    *lhs_and_rhs_tuple)
+
+    # TODO (Israt): Add support for 'max', 'min', 'mean' in heterograph
 
     # divide in degrees for mean reducer.
     if reduce_op == 'mean':
+        if g._graph.number_of_etypes() > 1:
+            raise NotImplementedError("Reduce op 'mean' is not supported in "
+                                      "the new heterograph API. Use multi_update_all().")
```

Comment:
Users will not understand what this *new* API is referring to in the future. `"Cannot set both intra-type and inter-type reduce operators as 'mean' using update_all. Please use multi_update_all instead."`. That being said, such error should be raised inside the `update_all` function instead of here. The reason is that `gspmm` might be invoked by other high-level APIs than `update_all` and this error message will be out-of-context.